tests/LinAlg/vectorTest.hpp

```diff
@@ -1527,19 +1537,151 @@ class VectorTests : public TestBase
     return reduceReturn(fail, &x);
   }
 
-protected:
-  // Interface to methods specific to vector implementation
-  virtual void setLocalElement(hiop::hiopVector* x, local_ordinal_type i, real_type val) = 0;
-  virtual real_type getLocalElement(const hiop::hiopVector* x, local_ordinal_type i) = 0;
-  virtual local_ordinal_type getLocalSize(const hiop::hiopVector* x) = 0;
-  virtual real_type* getLocalData(hiop::hiopVector* x) = 0;
-  virtual int verifyAnswer(hiop::hiopVector* x, real_type answer) = 0;
+  /// Method to set vector _x_ element _i_ to _value_.
+  void setLocalElement(hiop::hiopVector* x, local_ordinal_type i, real_type val)
+  {
+    x->copyFromDev();
+    real_type *xdat = x->local_data_host();
+    xdat[i] = val;
+    x->copyToDev();
+  }
+  /// Returns element _i_ of vector _x_.
+  real_type getLocalElement(const hiop::hiopVector* x, local_ordinal_type i)
+  {
+    x->copyFromDev();
+    return x->local_data_host_const()[i];
+  }
+
+  /// Returns pointer to local vector data
+  local_ordinal_type getLocalSize(const hiop::hiopVector* x)
+  {
+    return static_cast<local_ordinal_type>(x->get_local_size());
+  }
+
+  /// Returns pointer to local vector data
+  virtual real_type* getLocalData(hiop::hiopVector* x)
+  {
+    x->copyFromDev();
+    return x->local_data_host();
+  }
+
+  /// Checks if _local_ vector elements are set to `answer`.
+  int verifyAnswer(hiop::hiopVector* x, real_type answer)
+  {
+    x->copyFromDev();
+    const local_ordinal_type N = getLocalSize(x);
+    auto* xdata = static_cast<const real_type*>(getLocalData(x));
+    
+    int local_fail = 0;
+
+    for(local_ordinal_type i = 0; i < N; ++i)
+    {
+      if(!isEqual(xdata[i], answer))
+      {
+        ++local_fail;
+      }
+    }
+
+    return local_fail;
+  }
+
   virtual int verifyAnswer(
       hiop::hiopVector* x,
-      std::function<real_type(local_ordinal_type)> expect) = 0;
-  virtual bool reduceReturn(int failures, hiop::hiopVector* x) = 0;
-  virtual real_type* createLocalBuffer(local_ordinal_type N, real_type val) = 0;
-  virtual void deleteLocalBuffer(real_type* buffer) = 0;
+      std::function<real_type(local_ordinal_type)> expect)
+  {
+    x->copyFromDev();
+    const local_ordinal_type N = getLocalSize(x);
+    auto* xdata = static_cast<const real_type*>(getLocalData(x));
+    
+    int local_fail = 0;
+
+    for(local_ordinal_type i = 0; i < N; ++i)
+    {
+      if(!isEqual(xdata[i], expect(i)))
+      {
+        ++local_fail;
+      }
+    }
+
+    return local_fail;
+  }
+
+  /// If test fails on any rank set fail flag on all ranks
+  bool reduceReturn(int failures, hiop::hiopVector* x)
```

Comment:
This function is why I had to include `hiopVectorPar` and `hiopVectorRajaPar`. Since `get_mpi_comm()` is not a part of the `hiopVector` abstract interface, this seemed like the most reasonable solution.

Reply:
I'm a little worried about implementation-specific stuff creeping in. I think we should find a way around this and keep hiopVectorPar and hiopVectorRajaPar details out of tests/LinAlg/vectorTest.hpp at all costs

Reply:
I agree with this, but I did not want to make changes to the abstract interface without double-checking.

Reply:
@cnpetra Are you opposed to adding `get_mpi_comm` to hiopVector interface if `HIOP_USE_MPI=ON`?